### PR TITLE
bind indexed variables

### LINKED ISSUE #1785: How to bind indexed variables
### Description

Check out this app where I generate selectors dynamically:
image

```python
from taipy.gui import Gui
import taipy.gui.builder as tgb

questions = ["Are you alive ?", "aRe yOU sUre ?"]
answers = [""] * len(questions)


def selector(state):
    print(state.answers)


with tgb.Page() as page:
    for i, question in enumerate(questions):
        tgb.text(f"#### {question}", mode="md")
        tgb.selector(
            value="{answers[" + str(i) + "]}",
            lov=["Yes", "No"],
            on_change=selector,
        )


Gui(page).run()
```

How do I store the answers (the values selected)? Here, I try to store the answers in the `answers` list, but instead of assigning the selected value to `answers[0]`, it assigns it to `answers`, so I lose the other answers.

### Solution Proposed

There are multiple solutions but one should be implemented and documented:
- We could let the user bind a selector to an element of a list like I tried to do above
- We could let the user not bind the selector and define a var_name property. This way the user could store answers at the on_change(state, var_name, value) level

Currently, the workaround is using exec to create variables on the fly and it is not elegant:

```python
def selector(state, var_name, value):
    index = int(var_name[10:])
    state.answers[index] = value
    state.answers = state.answers
    print(state.answers)


with tgb.Page() as page:
    for i, question in enumerate(questions):
        tgb.text(f"#### {question}", mode="md")
        exec(f"selection_{i} = ''")
        tgb.selector(
            value="{selection_" + str(i) + "}",
            lov=["Yes", "No"],
            on_change=selector,
        )
```

### Impact of Solution

_No response_

### Additional Context

_No response_

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
resolves #1785

```
import taipy.gui.builder as tgb
from taipy.gui import Gui

questions = ["Are you alive ?", "aRe yOU sUre ?"]
answers = [""] * len(questions)


def selector(state):
    print(state.answers)


with tgb.Page() as page:
    for i, question in enumerate(questions):
        tgb.text(f"#### {question}", mode="md")
        tgb.selector(
            value=f"{{answers[{str(i)}]}}",
            lov=["Yes", "No"],
            on_change=selector,
        )


Gui(page).run(title="1785 bind indexed variables")

```

### CODE REVIEW

File: taipy/gui/utils/_evaluator.py [hunk 1]
```diff
@@ -293,15 +295,23 @@ def re_evaluate_expr(self, gui: Gui, var_name: str) -> t.Set[str]:
...
-                for v in temp_expr_var_map.values():
-                    var_name = v
+                var_name = next(iter(temp_expr_var_map.values()), var_name)
                 # construct correct var_path to reassign values
                 var_name_full, _ = _variable_decode(expr_original)
                 var_name_full = var_name_full.split(".")
                 var_name_full[0] = var_name
                 var_name_full = ".".join(var_name_full)
-                _setscopeattr_drill(gui, var_name_full, _getscopeattr(gui, var_name_original))
+                if index_in_array > - 1:
```
Comment (namnguyen20999) [inline taipy/gui/utils/_evaluator.py]: I think >= 0 would be clearer & intuitive

Comment (FredLL-Avaiga) [reply taipy/gui/utils/_evaluator.py]: how would that be different ?

File: taipy/gui/utils/_evaluator.py [hunk 2]
```diff
@@ -293,15 +295,23 @@ def re_evaluate_expr(self, gui: Gui, var_name: str) -> t.Set[str]:
             expr_original = self.__hash_to_expr[var_name]
             temp_expr_var_map = self.__expr_to_var_map[expr_original]
             if len(temp_expr_var_map) <= 1:
+                index_in_array = int(m[0]) if (m := _Evaluator.__IS_ARRAY_EXPR_RE.findall(expr_original)) else -1
                 # since this is an edge case --> only 1 item in the dict and that item is the original var
-                for v in temp_expr_var_map.values():
-                    var_name = v
+                var_name = next(iter(temp_expr_var_map.values()), var_name)
```
Comment (namnguyen20999) [inline taipy/gui/utils/_evaluator.py:303]: maybe for better clarity: `next(iter(temp_expr_var_map.values()), None) if temp_expr_var_map else var_name`

**Review by namnguyen20999 — DISMISSED** [review]

File: taipy/gui/utils/_evaluator.py (hunk 2 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_evaluator.py:303]: not sure it is clearer

File: taipy/gui/utils/_evaluator.py (hunk 1 above)
Comment (FabienLelaquais) [reply taipy/gui/utils/_evaluator.py]: I kind of agree with @namnguyen20999... >= 0 is toe me more intuitive than > -1.... it's not just me

**Review by FabienLelaquais — DISMISSED** [review]
Niice

File: taipy/gui/utils/_evaluator.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_evaluator.py]: everybody agrees even though it's 1 character less :-)

File: taipy/gui/utils/_evaluator.py [hunk 3]
```diff
@@ -108,7 +109,9 @@ def _analyze_expression(
                 if isinstance(node, ast.Name):
                     var_name = node.id.split(sep=".")[0]
                     if var_name in builtin_vars:
-                        _warn(f"Variable '{var_name}' cannot be used in Taipy expressions as its name collides with a Python built-in identifier.")  # noqa: E501
+                        _warn(
+                            f"Variable '{var_name}' cannot be used in Taipy expressions as its name collides with a Python built-in identifier."
+                        )  # noqa: E501
```
Comment (dinhlongviolin1) [inline taipy/gui/utils/_evaluator.py]: Can remove the ignore here

File: taipy/gui/utils/_evaluator.py [hunk 4]
```diff
@@ -198,6 +201,7 @@ def __get_holder_hash(holder: t.Type[_TaipyBase], expr_hash: str) -> str:
         return f"{holder.get_hash()}_{_get_client_var_name(expr_hash)}"
 
     def __evaluate_holder(self, gui: Gui, holder: t.Type[_TaipyBase], expr: str) -> t.Optional[_TaipyBase]:
+        expr_hash = ""
```
Comment (dinhlongviolin1) [inline taipy/gui/utils/_evaluator.py:205]: Why do we need the empty declaration here

**Review by dinhlongviolin1 — DISMISSED** [review]
Lgtm

**Review by namnguyen20999 — DISMISSED** [review]

File: taipy/gui/utils/_evaluator.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_evaluator.py]: if I cut the string yes

File: taipy/gui/utils/_evaluator.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/_evaluator.py:205]: it is used in the except block

**Review by dinhlongviolin1 — APPROVED** [review]

File: taipy/gui/utils/_evaluator.py [hunk 5]
```diff
@@ -303,14 +306,14 @@ def re_evaluate_expr(self, gui: Gui, var_name: str) -> t.Set[str]: # noqa C901
                 var_name_full = var_name_full.split(".")
                 var_name_full[0] = var_name
                 var_name_full = ".".join(var_name_full)
-                if index_in_array > - 1:
-                    array_val =  _getscopeattr(gui, var_name)
+                if index_in_array >= 0:
```
Comment (FabienLelaquais) [inline taipy/gui/utils/_evaluator.py:309]: Sooo better!

**Review by FabienLelaquais — APPROVED** [review]

**Review by namnguyen20999 — APPROVED** [review]
lgtm

**Pull request merged**